### ISSUE TITLE
Update OTEL trace_id value from a request header

### ISSUE THREAD

**Issue by Samjin** [apollographql/router#5056]
**Is your feature request related to a problem? Please describe.**

There is an existing infrastructure that utilizes a dedicated request header for trace ID. In Router, it'd make sense to use that as the value of OTEL `trace_id` for all logging, metrics, and span tags. This could reduce multiple trace ids labeled everywhere and also helps to correlate request logs from apollo-router, otherwise user may have to lookup 2 different trace ids to find a request logs.  

i.e. 
```
{ [-]
   dd.trace_id: 7188631615460499999
   fetch_error: SubrequestHttpError { status_code: None, service: "xxx", reason: "error trying to connect: Connection reset by peer (os error 104)" }
   filename: /usr/local/cargo/git/checkouts/router-18de2f110854a520/155375b/apollo-router/src/services/subgraph_service.rs
   level: ERROR
   span_id: fc8f2f9c6b29999
   trace_id: 000000000000000063c3265e04aa1b0f
   ...
}
```

As you can see above, the trace_id is not the one our infrastructure use, but generated by apollo router. 

i.e. we use `trace_id` for Splunk and Datadog in current Node gateway, and would like to continue to use same names in Router. If `display_trace_id: true` is enabled from logging, there maybe be two trace_ids that could potentially overwrite each other if located at same level(I have not check if this is the case today but maybe possible?).  

**Describe the solution you'd like**

Have an option to override default `trace_id` value from a request header's UUID like value. This should update default trace_id value everywhere, including span attributes, logging, metrics etc. 

trace id header will be created in router_service if it doesn't exist, this created header should be able to propagate as trace_id as well

**Comment by abernix** [comment]
The router would only originate a trace Id when it doesn't get one from the headers.  Do you have tracing [propagation](https://www.apollographql.com/docs/router/configuration/telemetry/exporters/tracing/overview/#propagation) setup or any exporters enabled?  Can you share more of your configuration?

**Comment by Samjin** [comment]
> The router would only originate a trace Id when it doesn't get one from the headers. Do you have tracing [propagation](https://www.apollographql.com/docs/router/configuration/telemetry/exporters/tracing/overview/#propagation) setup or any exporters enabled? Can you share more of your configuration?

I tried it and left comment here. https://github.com/apollographql/router/pull/4823#issuecomment-2085656172
```
          datadog:
            enabled: true
            enable_span_mapping: true
            batch_processor:
              ...
            endpoint: "xx"
          propagation:
            datadog: true
            request:
              header_name: trace-id
```
It either updated the our request header by trimming off the `-`, or replaced with `trace_id` value. Our downstream received  value like `traceid=28e9fb547e465690f288cdb96194d119`

**Comment by Samjin** [comment]
I still get a different trace_id than trace-id header used above. The log is from `subgraph_service`. 

<img width="594" alt="image" src="https://github.com/apollographql/router/assets/6115094/dd3eb99c-2ffb-4787-8cb7-3f9f09ea8513">

And, subgraph would receive this error(invalid UUID) from trace-id header. So somewhere overwrite trace-id header. 
<img width="650" alt="image" src="https://github.com/apollographql/router/assets/6115094/26776080-b162-4027-a1d8-68fc78307d26">

Also the following trace ids are mutual exclusive. they override each over in span attrs. 
```
      router:
        attributes:
          trace_id: true
          trace-id:
            request_header: trace-id
```

**Comment by Samjin** [comment]
In debug mode, I get following error
```
  "message": "cannot generate custom trace_id: invalid digit found in string",
  "target": "apollo_router::plugins::telemetry",
```

looks like trace_id is recreated because it couldn't parse UUID header.

**Comment by abernix** [comment]
What is your subgraph server implemented in and what telemetry mechanism (SDK, format, etc.) are you using in it?

**Comment by BrynCooke** [comment]
@Samjin I think that this PR will likely fix things for you: https://github.com/apollographql/router/pull/5071
We are targeting this for the next release.

**Comment by Samjin** [comment]
<img width="956" alt="image" src="https://github.com/apollographql/router/assets/6115094/1cf5a846-3698-4ee4-ae77-b2a1b4715ae1">

<img width="595" alt="image" src="https://github.com/apollographql/router/assets/6115094/779c542d-5fdc-48bd-a848-6162aec8c192">

I'm using latest dev, both custom and apollo-router events are logged with `trace_id`, which is now using original UUID but  without dashes, but the id is only added in `deprecate` spans mode. `trace_id` doesn't appear in`spec_compliant` logs. 

That said, even if it is successfully added in `spec_compliant` mode, it's still a different trace id than the UUID that up/downstream services use. So I'm not completely sure how removing `-` from UIUD would work. 

I think to fix this temporarily, the easy way maybe just allow user to have a req header selector to create custom attribute for all logs, then they can just disable the default trace_id. Doing this way may require user to add their own trace id to their custom events.

**Comment by Samjin** [comment]
> What is your subgraph server implemented in and what telemetry mechanism (SDK, format, etc.) are you using in it?

They use opentracing in SDK for datadog. It looks like trace context use trace-id https://github.com/opentracing/specification/blob/master/rfc/trace_identifiers.md#trace-context-http-headers

But still Im not completely sure they are related, it'd be great if anyone know how they work together.

**Comment by smyrick** [comment]
I think we can close this issue. We already allow parsing a `trace-id` from any given header but to remain OpenTelemetry compliant, that header must a 1[6-btye random generated string](https://opentelemetry.io/docs/specs/otel/overview/#spancontext) and not something that includes dashes. We are even now [striping dashes](https://github.com/apollographql/router/pull/5071) from any header value if we see them to propagate the correct value.

If you have clients passing in trace ids with dashes those are not valid traces id and you should fix those at the source


Now that being said, it would be helpful if you could include some company specific ids in you logs as standard attributes and we have documented that separately here: https://github.com/apollographql/router/issues/5212

**Comment by Samjin** [comment]
To give a summary, these are mentioned in the thread. #1 is what this issue is asking for. For #2, should I open another issue separately or is this expected behavior?

1. propagate custom trace_id header
2. log injection feature is not appending `trace_id` and `span_id` to user's custom event in `spec_compliant` mode like `deprecated` mode does.

**Comment by smyrick** [comment]
@Samjin If the JSON log formatter had the option to toggle specific attributes, would that work? https://github.com/apollographql/router/issues/5212#issuecomment-2148120150

**Comment by BrynCooke** [comment]
I believe this was fixed in #5802. Please reopen if this is not the case.
